### PR TITLE
[ETH-245] TokenGated Rooms: ERC721

### PR DESCRIPTION
**Pull request by bearni95**
- implemented deployment, self-promotion and pinning of ERC721-gated rooms
- room properties modal now includes info about the ERC721 token (name, symbol, address, tokenId, json metadata)
- extended the ERC20 sagas into accommodating multiple token standards

**Test ERC721 Token:**
A worthless ERC721 token has been deployed on Polygon to test this functionality. It's address is `0xBCB999e8340327B40293bf785d8913587879e0F6` and 1 token with id `1234567890` has been minted to Streamr Dev 5 (`0x00dE714Cbad811aF322F539A043EC71eaB7Fa3a5`)

### CODE REVIEW

File: src/features/tokenGatedRooms/sagas/registerPolicy.saga.ts [hunk 1]
```diff
@@ -25,40 +26,73 @@ function createInformationalToast(message: string, toastId?: string | number) {
...
+    let policyAddress: Address
+    if (tokenType.standard === TokenTypes.ERC20.standard) {
+        policyAddress = await factory.erc20TokensToJoinPolicies(tokenAddress, roomId)
+    } else if (tokenType.standard === TokenTypes.ERC721.standard) {
+        policyAddress = await factory.erc721TokensToJoinPolicies(tokenAddress, tokenId, roomId)
+    } else {
+        throw new Error('Unknown token type')
+    }
 
     if (policyAddress === '0x0000000000000000000000000000000000000000') {
```
Comment (mondoreale) [inline src/features/tokenGatedRooms/sagas/registerPolicy.saga.ts:44]: It's possible, theoretically, that this ends up being an infinite loop.

**Review by mondoreale — COMMENTED** [review]
Leaving a couple of comments here. Let's address them. Looks pretty nice so far!

File: src/components/modals/AddRoomModal.tsx [hunk 2]
```diff
@@ -245,8 +246,10 @@ export default function AddRoomModal({ setOpen, ...props }: ModalProps) {
                                 <Label htmlFor="tokenId">Token ID</Label>
                                 <TextField
                                     id="tokenId"
-                                    value={tokenId}
-                                    onChange={(e) => void setTokenId(parseInt(e.target.value))}
+                                    value={tokenId.toString()}
+                                    onChange={(e) =>
+                                        void setTokenId(BigNumber.from(e.target.value))
```
[the remark this answers is not in the document]
Comment (bearni95) [reply src/components/modals/AddRoomModal.tsx:251]: I'd assume the initial value setting prevents the input from ever being empty:
https://github.com/streamr-dev/chat/blob/cc36b7df61e7769b8ee11c1bb2791e996ad50ae2/src/components/modals/AddRoomModal.tsx#L53

File: src/features/room/sagas/pin.saga.ts [hunk 3]
```diff
@@ -33,18 +33,18 @@ function* pinRemote(
...
-    const { createdAt, createdBy, tokenAddress } = stream.extensions['thechat.eth']
+    const { createdAt, createdBy } = stream.extensions['thechat.eth']
 
     const isTokenGated = isTokenGatedRoom(stream)
 
-    if (isTokenGated && tokenAddress) {
-        yield put(
-            TokenGatedRoomAction.joinERC20({
+    if (isTokenGated) {
+        yield putResolve(
```
[the remark this answers is not in the document]
Comment (bearni95) [reply src/features/room/sagas/pin.saga.ts:41]: Originally part of a way to make the process halt, I see now that `putResolve` behaves just like `put`, so reverting to the former

[the remark this answers is not in the document]
Comment (mondoreale) [reply src/features/room/sagas/pin.saga.ts]: Yeah, `putResolve` is new to me. Found an example which illustrates the difference.
https://codesandbox.io/s/redux-saga-example-x77jb?file=/sagas/index.js

**Pull request closed without merging**